gulpfile.js

```diff
@@ -23,9 +29,84 @@ gulp.task('css', function () {
     .pipe(livereload());
 });
 
+function build_main(name, callback) {
+  // build main.min.js for a given application name
+  // run in a subprocess to allow parallel builds
+  // clone requirejs config
+  var p = fork('./build-main.js', [name]);
+  p.on('exit', function (code, status) {
+    if (code) {
+      callback(new Error("Build failed"));
+    } else {
+      callback();
+    }
+  });
+  return;
+}
+
+// build notebook-js, edit-js, etc. tasks
+// which enables parallelism
+var apps = ['notebook', 'tree', 'edit', 'terminal', 'auth'];
+
+apps.map(function (name) {
+  gulp.task(name + '-js', function (finish) {
+    var s = path.join('notebook', 'static');
+    var src = path.join(s, name, 'js', 'main.js');
+    var rel_dest = path.join(name, 'js', 'main.min.js');
+    var dest = path.join(s, rel_dest);
+    
+    var sources = [
+      path.join(s, name, 'js', '*.js'),
+      path.join(s, "base", 'js', '*.js'),
+      path.join(s, "auth", 'js', '*.js'),
+      path.join(s, "services", 'config.js'),
+    ];
+    
+    // for required_components
+    if (name === 'notebook') {
+      sources.push(path.join(s, "services", '**', '*.js'));
+    }
+    
+    fs.readdirSync(path.join(s, 'components')).map(function (c) {
+      if (c !== 'MathJax') {
+        // skip MathJax because it has tons of files and makes everything super slow
+        sources.push(path.join(s, 'components', c, '**', '*.js'));
+      }
+    });
+    
+    gulp.src(sources, {base: s})
+    .pipe(newer(dest))
```

Comment:
Wont this fail if the changed file is one of the dependencies?  i.e.  main.js imports notebook.js, and I change notebook.js, main.js's modified timestamp doesn't change, so re-optimization wont happen because it will think nothing has changed.